### PR TITLE
Add state boundary conditions

### LINKED ISSUE #98: Add state boundary conditions
**Is your feature request related to a problem? Please describe.**
Currently, boundary conditions in the land model are expressed in terms of fluxes (i.e. heat and/or water flux). We would like to be able to also specify state boundary conditions (i.e. temperature and/or water content). Any combination of different boundary conditions should be able to be used, in particular the case of one flux and one state.

**Describe the solution you'd like**
Create a new subtype of `AbstractBoundaryConditions` for state boundary conditions
Add new methods of `boundary_fluxes` to be able to handle any combination of 1 or 2 flux and/or state boundary conditions

**Describe alternatives you've considered**

**Additional context**
Linked PR: #103

### PR DESCRIPTION
**Pull request by juliasloan25**
## Purpose 

Extend boundary conditions (BCs) in LSM models to be able to include state BCs in addition to flux BCs (and potentially other new BC types).
Closes #98 

## To-do

- Implement and test in RRE+heat case

## Content

- Add StateBC type
- Perform calculations to convert state BCs to flux BCs
- Separate top/bottom BC calculations
- Implement and test in RRE-only case
- Implement and test in RRE/heat joint case

### CODE REVIEW

File: src/Soil/Soil.jl [hunk 1]
```diff
@@ -75,24 +75,44 @@ import ClimaLSM:
...
+"""
+    AbstractBC{FT <: AbstractFloat}
+
 An abstract type for types of boundary conditions, which will include
 prescribed functions of space and time as Dirichlet conditions or
 Neumann conditions, in addition to other 
 convenient soil-specific conditions, like free drainage.
 """
-abstract type AbstractSoilBoundaryConditions{FT <: AbstractFloat} end
+abstract type AbstractBC{FT <: AbstractFloat} end
```
Comment (kmdeck) [inline src/Soil/Soil.jl:115]: do we need AbstractBC and AbstractSoilBoundaryConditions?

File: Project.toml [hunk 2]
```diff
@@ -6,8 +6,10 @@ version = "0.1.0"
 [deps]
 CLIMAParameters = "6eacf6c3-8458-43b9-ae03-caf5306d3d53"
 ClimaCore = "d414da3d-4745-48bb-8d80-42e94e092884"
+DiffEqCallbacks = "459566f4-90b8-5000-8ac3-15dfb0a30def"
```
Comment (kmdeck) [inline Project.toml:9]: I believe these are just in the test environment for now (the ClimaLSM package is in some sense agnostic to timestepper). we can discuss if this makes sense

File: docs/src/APIs/Soil.md [hunk 3]
```diff
@@ -45,8 +45,15 @@ ClimaLSM.Soil.EnergyHydrologyParameters
...
-ClimaLSM.Soil.boundary_fluxes
+ClimaLSM.Soil.AbstractBC
+ClimaLSM.Soil.StateBC
 ClimaLSM.Soil.FluxBC
+ClimaLSM.Soil.AbstractSoilBoundaryConditions
 ClimaLSM.Soil.AbstractSoilSource
 ClimaLSM.Soil.source!
+ClimaLSM.Soil.RootExtraction
+ClimaLSM.Soil.AbstractBoundary
+ClimaLSM.Soil.TopBoundary
```
Comment (kmdeck) [inline docs/src/APIs/Soil.md:56]: could we call it BottomBoundary for clarity?

File: src/Soil/Soil.jl [hunk 4]
```diff
@@ -165,30 +192,208 @@ function dss!(
...
+
+"""
+    get_level(field, idx::Int64
+
+A simple helper function to extract the value stored at the level specififed
+by idx. Note that this currently works for fields representing columns, but the index
+of 1 may need to change to support larger fields.
+"""
+# TODO change this to support non-column fields
+get_level(field, idx::Int64) = parent(Fields.level(field, idx))[1]
```
Comment (kmdeck) [inline src/Soil/Soil.jl:304]: yes, let's remove the parent() and indexing here

File: src/Soil/Soil.jl [hunk 5]
```diff
@@ -165,30 +192,208 @@ function dss!(
...
+    return top_Δz, bottom_Δz
+end
+
+"""
+    rre_flux_calc(K, Ψ_bc, Ψ_c, Δz)
+
+Calculates the water flux at a boundary using the given parameters.
+"""
+function rre_flux_calc(K, Ψ_bc, Ψ_c, Δz)
+    return @. -K * ((Ψ_bc - Ψ_c) + (Δz / 2)) / (Δz / 2)
```
Comment (kmdeck) [inline src/Soil/Soil.jl:326]: just for clarity, can we pass in dz/2, but leave it as dz here?

File: src/Soil/Soil.jl [hunk 6]
```diff
@@ -165,30 +192,208 @@ function dss!(
...
+    T_rre_top <: AbstractBC{FT},
+    T_heat_top <: AbstractBC{FT},
+    T_rre_bot <: AbstractBC{FT},
+    T_heat_bot <: AbstractBC{FT},
+} <: AbstractSoilBoundaryConditions{FT}
+    bc_top::Tuple{T_rre_top, T_heat_top}
+    bc_bot::Tuple{T_rre_bot, T_heat_bot}
+end
+
+function RREHeatBoundaryConditions{FT}(
```
Comment (kmdeck) [inline src/Soil/Soil.jl:278]: do we need these type containers? Or could we use named Tuples? (heat = (top = , bot = ), water = (top = , bot = ))

File: src/Soil/Soil.jl [hunk 7]
```diff
@@ -165,30 +192,208 @@ function dss!(
...
+    # Approximate K_bc ≈ K_c, Ψ_bc ≈ Ψ_c (center closest to the boundary)
+    K_c = Fields.level(p.soil.K, 1)
+    Ψ_c = Fields.level(p.soil.ψ, 1)
+
+    # Calculate pressure head using boundary condition
+    @unpack vg_α, vg_n, vg_m, θ_r, ν, S_s = params
+    Ψ_bc = @. pressure_head(vg_α, vg_n, vg_m, θ_r, rre_bc.bc, ν, S_s)
+
+    # TODO how do I return negative value of Field?
+    return -rre_flux_calc(K_c, Ψ_bc, Ψ_c, Δz_bot)
```
Comment (kmdeck) [inline src/Soil/Soil.jl:376]: let's double check this for the bottom boundary!

File: src/Soil/Soil.jl [hunk 8]
```diff
@@ -165,30 +192,208 @@ function dss!(
...
+end
+
+# Convert heat state to flux at bottom boundary
+function boundary_flux(heat_bc::StateBC, ::BotBoundary, Δz_bot, p)
+    # Approximate κ_bc ≈ κ_c (center closest to the boundary)
+    T_c = Fields.level(p.soil.T, 1)
+    κ_c = Fields.level(p.soil.κ, 1)
+
+    # TODO how do I return negative value of Field?
+    return -heat_flux_calc(κ_c, heat_bc.bc, T_c, Δz_top)
```
Comment (kmdeck) [inline src/Soil/Soil.jl:396]: let's double check this for the bottom boundary!

**Comment by juliasloan25** [comment]
bors try
